Implement login and logout

tfidfwastaken opened the pull request:
Closes [sc-135].

We store sessions in memory for the time being. Adding a persistent external store is possible without change to this code by implementing Ring's [SessionStore protocol](https://github.com/ring-clojure/ring/wiki/Sessions#session-stores). This is a task reserved for later (cf. [ADR-001](https://github.com/nilenso/bugle-forms/blob/4cf9e36edc5580edc9159e10b3d0678b671dc28c/doc/arch/adr-001-sessions-for-auth.md)).

tfidfwastaken committed:
handlers: change error-redirect util

303 is the correct status message for the redirect utility's purpose,
which is to give an indirect response to the user. We also change its
name to flash-redirect as this may not only be used for failure modes.

tfidfwastaken committed:
handlers: add dashboard stub and redirects

We add a stub page for dashboard and redirect users to the appropriate
page based on whether they are in a session or not.

tfidfwastaken committed:
views: display a login form

tfidfwastaken committed:
user: implement login and logout

We use server-side sessions to implement login and logout. Ring's
session middleware is used.

tfidfwastaken committed:
user tests: test login and logout functionality

shafeeq commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -8,17 +8,33 @@
+  "Create a session on a successful login."
+  [{:keys [form-params session]}]
+  (if (:user session)
+    (response/redirect "/dashboard" :see-other)
```
Use a middleware that can do the redirect to `/dashboard` if the session already exists. That way we can avoid checking for session in each handler.

tfidfwastaken replied:
Makes sense. 

Related: would it make sense to have a middleware for the converse of this—ie, trying to access the dashboard _without_ a session should redirect to `/login`. This case will be more common than the one you are talking about—there will be more privileged routes than unprivileged ones down the line.

Thinking off-the-cuff right now, but I think these two cases should be unified somehow in the same middleware.

shafeeq replied:
You're right, we'll need to handle redirects to login as well. These two can be different middlewares that we can selectively apply to handlers

shafeeq commented on src/bugle_forms/model/user.clj:
```diff
@@ -24,3 +28,27 @@
+            ["select name, password from user_account where email = ?" email])]
+    (if (:valid (hashers/verify given-pwd pwd))
+      {:id (generate-session-id)
+       :name name}
```
We should store some user identifier in session for use later. `name` is not a good identifier becuse it's not unique.

tfidfwastaken replied:
Okay. I didn't intend to use `:name` as an identifier, rather it was meant as data to personalize messages to a logged in user ("Hi, `:name` 👋 "). I didn't actually think of having including the user's unique identifier at all in the session, but I'll eventually need it for sure. Thanks for calling this out.

shafeeq commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -8,17 +8,33 @@
+                        {:title "Sign up" :flash flash}
+                        user-views/signup))))
+
+(defn create-session
```
The name `create-session` is a bit confusing. It seems to suggest that the fn mostly takes care of creating session based on the current implementation, but here it's actually the handler for the POST request for login, which might or might not create the session based on credential verification.

Maybe rename this to `login` and the other GET handler to `login-form`?

shafeeq commented on test/bugle_forms/handlers/user_test.clj:
```diff
@@ -28,5 +28,79 @@
+                               :password "hunter2"}
+            login-request {:form-params login-form-params}
+            response (sut/create-session login-request)]
+        (is (= 303 (:status response)))
```
Also test for absence of session?

shafeeq commented on test/bugle_forms/model/user_test.clj:
```diff
@@ -15,3 +20,25 @@
+    (testing "Existing user is assigned a valid session ID"
+      (let [login-params (select-keys user-data [:email :password])
+            auth-response (user/authenticate login-params)]
+        (is (string? (:id auth-response)))
```
Also test for the id being not-empty, or a fixed length if it is expected.

shafeeq reviewed (changes requested):
Looks good, requested some minor changes

tfidfwastaken committed:
refactor routing and validation middleware

What has changed:
- Routing is now having a layer of indirection—we dispatch on the
  keywords returned by a route and process it according to the handler
  specification for the given keyword.
- Redirects are now handled in middleware, which is utilized in the
  handler dispatch layer.
- Tests are moved around as a result of these changes.
- Extra validation has been added to login form.
- Session data now contains the user's UUID.

shafeeq commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -6,25 +6,48 @@
+  "Log the user out; deletes session."
+  [{{:keys [user]} :session}]
+  (-> (util/flash-redirect "/" "Logged out successfully.")
+      (assoc :session nil)))
```
Use `dissoc`

tfidfwastaken replied:
This actually won't work in deleting the session. The assoc of nil is information for the Ring `wrap-session` middleware to remove the session. There anyway did not exist a `:session` key in the response map, so there is nothing to `dissoc`, so using that will break logout.

shafeeq commented on src/bugle_forms/middleware.clj:
```diff
@@ -24,3 +28,51 @@
+  If the `:spec` option is given, the handler is called only when the `:field`
+  value in the request matches the spec, else a 400 status response is
+  returned."
+  [handler {:keys [spec field]}]
```
Is field a necessary argument? Can we always just read `params` from the request map?

tfidfwastaken replied:
In my opinion, it is useful because we might want to validate our spec on various possible request fields. Eg:
- Validate form submission fields only, ie, `:form-params`
- Validate query strings only, ie, `:query-params`
- Validate some combination of both, ie, `:params`
- (more esoteric?) Validate some other request fields, eg, `:scheme` should be `https` or `X-csrf-token` header exists.

Granted, I only ever validate form-params in my application so far, so this may have been a case of premature abstraction.

shafeeq commented on test/bugle_forms/handlers/user_test.clj:
```diff
@@ -28,5 +28,41 @@
+            login-request {:form-params login-form-params}
+            response (sut/login login-request)]
+        (is (= 303 (:status response)))
+        (is (nil? (get-in response [:headers "Set-Cookie"])))
```
Since the request is made directly to the user handler, the response here won't have any headers for setting the session cookie - it is added when it passes through the session middleware in the middleware chain.

At this point, the tests for session should be just based on `:session` in response map

tfidfwastaken replied:
Got it, thanks for catching this!

shafeeq reviewed (commented):
Good job on extracting the middleware usage to a declarative form. The tests for it looks good, but they all use generated handlers. I would suggest adding a test that simulates a request all the way through the middleware chain as well - just to verify that they are integrated correctly.
